example/gmt/sphere_preprocessing.c

```diff
@@ -0,0 +1,525 @@
+      }
+
+      /* Record segments given by the computed endpoints */
+      for (p4est_topidx_t tree = 0; tree < 6; tree++) {
+        if (endpoints_count[tree] == 0) {
+          /* The geodesic does not cross this cube face */
+          continue;
+        }
+        if (endpoints_count[tree] == 1) {
+          /* The geodesic has an endpoint on the edge of this face */
+          /* This is an edge case and probably never happens */
+          /* Set xyz1 = xyz2 to this point */
+          xyz1[0] = xyz2[0] = endpoints[tree][0][0];
+          xyz1[1] = xyz2[1] = endpoints[tree][0][1];
+          xyz1[2] = xyz2[2] = endpoints[tree][0][2];
+        }
+        if (endpoints_count[tree] == 2) {
+          /* The geodesic has a generic segment on this face */
+          /* Set xyz1 and xyz2 to computed endpoints */
+          xyz1[0] = endpoints[tree][0][0];
+          xyz1[1] = endpoints[tree][0][1];
+          xyz1[2] = endpoints[tree][0][2];
+          xyz2[0] = endpoints[tree][1][0];
+          xyz2[1] = endpoints[tree][1][1];
+          xyz2[2] = endpoints[tree][1][2];
+        }
+
+        /* Convert to tree-local coordinates */
+        p4est_geometry_cubed_Y (xyz1, rst1, tree);
+        p4est_geometry_cubed_Y (xyz2, rst2, tree);
+
+        /* Store geodesic segment */
+        geodesics[n_geodesics].which_tree = tree;
+        geodesics[n_geodesics].p1x = rst1[0];
+        geodesics[n_geodesics].p1y = rst1[1];
+        geodesics[n_geodesics].p2x = rst2[0];
+        geodesics[n_geodesics].p2y = rst2[1];
+
+        n_geodesics++;
+      }
+    }
+  }
+
+  /* Free extra capacity */
+  geodesics =
+    P4EST_REALLOC (geodesics, p4est_gmt_sphere_geoseg_t, n_geodesics);
+
+  *n_geodesics_out = n_geodesics;
+  *geodesics_out = geodesics;
+
+  return 0;
+}
+
+/** Load geodesics from input csv, convert to Cartesian coordinates, split into
+ *  segments corresponding to the trees in the cubed connectivity, then write
+ *  to array of type p4est_gmt_sphere_geoseg_t.
+*/
+int
+main (int argc, char **argv)
+{
+  int                 progerr;
+  const char         *usage;
+  FILE               *output = NULL;
+  FILE               *input = NULL;
+  p4est_gmt_sphere_geoseg_t *geodesics;
+  size_t              n_geodesics;
+  size_t              nwritten;
+
+  usage = "Arguments: <input.csv>\n";
+  progerr = 0;
+
+  if (argc != 2) {
+    P4EST_GLOBAL_LERROR ("Incorrect number of arguments\n");
+    P4EST_GLOBAL_LERROR (usage);
+    progerr = 1;
+  }
+
+  if (!progerr) {
+    /* open input file for reading */
+    input = fopen (argv[1], "r");
+    if (input == NULL) {
+      P4EST_GLOBAL_LERROR ("Could not find input file\n");
+      P4EST_GLOBAL_LERROR (usage);
+      progerr = 1;
+    }
+  }
+
+  /* split geodesics into segments */
+  if (!progerr) {
+    progerr = compute_geodesic_splits (&n_geodesics, &geodesics, input);
+  }
+
+  /* close input file */
+  if (input != NULL) {
+    fclose (input);
```

Comment:
Checking the return of every fclose will be nice, even though we'd only print a message since progerr is already true.

Reply:
It's a good thing not checking for progerr here and closing depending exactly when required.